### PR TITLE
"swap followed by measure" optimization pass

### PR DESCRIPTION
**Pull request by 1ucian0**
This PR goes on top of https://github.com/Qiskit/qiskit-terra/pull/1889

Removes the swap before the measurement.

### CODE REVIEW

**Comment by 1ucian0** [comment]
on hold until #1889 gets merged.

**Comment by 1ucian0** [comment]
PR https://github.com/Qiskit/qiskit-terra/pull/1976 merged. removing `on hold`.

File: qiskit/transpiler/passes/optimize_swap_before_measure.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                    node = dag.multi_graph.node[successor]
+                    if node['type'] == 'op' and node['op'].name == 'measure':
+                        # replace measure node with a new one, where qargs is set with the "other"
+                        # swap qarg.
+                        dag._remove_op_node(successor)
+                        old_measure_qarg = node['qargs'][0]
+                        new_measure_qarg = swap_qargs[swap_qargs.index(old_measure_qarg) - 1]
+                        measure_layer.apply_operation_back(
+                            Measure(qubit=new_measure_qarg, bit=node['cargs'][0]))
+                dag.extend_back(measure_layer)
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passes/optimize_swap_before_measure.py]: That created problems in the pass https://github.com/Qiskit/qiskit-terra/pull/2005.

File: qiskit/transpiler/passes/optimize_swap_before_measure.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                for qreg in dag.qregs.values():
+                    measure_layer.add_qreg(qreg)
+                for creg in dag.cregs.values():
+                    measure_layer.add_creg(creg)
+                for successor in dag.successors(swap):
+                    node = dag.multi_graph.node[successor]
+                    if node['type'] == 'op' and node['op'].name == 'measure':
+                        # replace measure node with a new one, where qargs is set with the "other"
+                        # swap qarg.
+                        dag._remove_op_node(successor)
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passes/optimize_swap_before_measure.py]: done in https://github.com/Qiskit/qiskit-terra/pull/2024

[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passes/optimize_swap_before_measure.py]: (merged to this branch in 1951d604036c7b7c7ced48856c46c087b56853a2)

**Comment by 1ucian0** [comment]
> this would cause a SWAP occurring at the end of a circuit to also be removed. But this pass should only remove SWAPs that are explicitly followed by measure. you should remove node['type'] == 'out'

Are you sure about this? Take the example from Figure 11 in the QASM specification (Example of a quantum ripple-carry adder from [29]. This circuit prepares a = 1, b = 15 and computes the sum into b with an output carry cout[0]). If there are swaps between a and b, they wont be optimized under your approach. But, why not?

File: qiskit/transpiler/passes/optimize_swap_before_measure.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class OptimizeSwapBeforeMeasure(TransformationPass):
+    """Remove the swaps followed measurement (and adapts the measurement)"""
+
+    def run(self, dag):
+        """Return a new circuit that has been optimized."""
+        swaps = dag.op_nodes(SwapGate)
+        for swap in swaps:
+            final_successor = []
+            for successor in dag.successors(swap):
+                node = dag.multi_graph.node[successor]
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passes/optimize_swap_before_measure.py]: fixed in dd26f919

**Comment by 1ucian0** [comment]
(This PR contains #2024)

**Comment by 1ucian0** [comment]
I still need a review for this :)

**Comment by ajavadia** [comment]
two files must be removed from this PR:
`qiskit/transpiler/passes/count_operations.py` and its test

**Comment by 1ucian0** [comment]
Fixed. I also moved the fixed point to check the DAG instead of counting ops.

**Comment by ajavadia** [comment]
Can you add a changelog line for this as well? Approved.

**Pull request merged**